New Post: CSRF - protecting ifixit.com

danielbeardsley opened the pull request:
Describes the problem, typical prevention methods, and our simple method.

danielbeardsley committed:
Post: CSRF

Work in progress

danielbeardsley committed:
Post: CSRF -- small typos

danielbeardsley committed:
Post: CSRF - a few more paragraphs

Add the entire csrf js implementation as an example

danielbeardsley committed:
Post: CSRF - grammar adjustments

and some sentence refactoring

danielbeardsley committed:
Post: CSRF - fix bug in js

call() takes individual arguments, apply() takes an array. I guess
we don't call submit() with arguments very often cause this was broken

danielbeardsley committed:
Post: CSRF - improve the paragraph about deployment

danielbeardsley committed:
Post: CSRF - more grammar fixes, yay proofreading

danielbeardsley committed:
Post: CSRF - replace js with embedded gist

cause that's cool.

danielbeardsley committed:
Post: CSRF - explain what the server does

Cause it wasn't super-clear before.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,89 @@
+## What does a CSRF attack allow?
+Cross-site request forgery allows an attacker
+to make arbitrary requests to a target site
+on behaf of another user,
```
behaf -> behalf

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,89 @@
+(maybe something important like deleting a Guide).
+Next, they create a similar form (via html or javascript) on their own site.
+When a user who is logged in on the target site visits the attackers site,
+it submits the form to the target site
```
`it` seems a little unclear here. I would simply change it to what exactly is submitting the form to the target site.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,89 @@
+We use javascript to dynamically set a cookie
+and add an `<input>` element to the form
+immediately before it's submitted.
+Because we just have to prove that we can alter or read cookies
```
You need a comma after cookies.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,89 @@
+and add an `<input>` element to the form
+immediately before it's submitted.
+Because we just have to prove that we can alter or read cookies
+we don't need to depend on the server-side for setting the cookie,
```
I would change this comma to a semicolon.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,89 @@
+## How we deployed this
+We designed it, tested it extensively, ran our test suite many times, then
+soft-deployed it, silently reporting CSRF failures to a database. Once we'd
+worked out a few issues and were confident that we'd covered all our bases we
```
You need a comma after bases.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,89 @@
+We designed it, tested it extensively, ran our test suite many times, then
+soft-deployed it, silently reporting CSRF failures to a database. Once we'd
+worked out a few issues and were confident that we'd covered all our bases we
+flipped the switch on enforcement and it's been working well since.
```
You need a comma after enforcement.

danielbeardsley committed:
Post: CSRF - fix moar grammar / spelling badness

Thanks to @brotatos for the suggestions.

danielbeardsley commented:
@brotatos Thanks for the proofread! I've made all of those corrections.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+
+We've come up with a minimalist method of
+protecting our site from cross site request forgery ([CSRF][csrf])
+and we'd like to share how we do it.
```
You need a comma before `and`.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+Cross-site request forgery allows an attacker
+to make arbitrary requests to a target site
+on behalf of another user,
+*with* their privleges but *without* their consent.
```
privleges -> privileges

I don't think you need the comma after `user` as well.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+At the very least, that's just more boiler-plate code
+that slows development progress.
+At the worst, you may forget to add the CSRF `<input>` tag
+to some seldom-used forms and they'll efectively be broken.
```
efectively -> effectively

danielbeardsley replied:
So, I'd never used spell checking in vim until _just_ now... it's awesome: `:set spell spelllang=en_us` and `:help spell`

All these errors prompted me to investigate.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+we don't need to depend on the server-side for setting the cookie;
+the client can do it just fine.
+We hook into forms by replacing the `form.submit()` function
+and listening for the `onsubmit` event.
```
listening -> listen

danielbeardsley replied:
I think `We do blah by replacing X and listening for Y` makes more sense that `We do blah by replacing X and listen to Y`

brotatos replied:
Sounds good to me; the sentence structure simply needs to stay parallel.

marczych replied:
I also thought it was wrong the first time I read it. It is either:
`We hook into X and listen for Y.` or
`We hook into blah by replacing X and listening for Y.`

I think the most obvious interpretation as you read it for the first time is `We hook into X and listening for Y.` Could we rewrite it to remove the ambiguity?

danielbeardsley replied:
I still don't understand, I must be reading my own sentence oddly. What is ambiguous about `We do blah by replacing X and listening for Y`?. The rewrite of `We hook into X and listen for Y.` leaves out the `by replacing the ...` part which explains one of the ways we "hook" into forms.

marczych replied:
The `X` in `We hook into X and listen for Y.` is `forms by replacing the 'form.submit' function`.

I've read the sentence too many times now so I can't decide whether it's fine or not.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+Here's the entire javascript implementation of our CSRF protection (Github [gist][gist]):
+{% gist 6060418 %}
+
+*Note: we use Mootools, but refactoring for no jQuery or no library would be
```
Comma splice.

brotatos commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+
+## How we deployed this
+We designed it, tested it extensively, ran our test suite many times,
+then soft-deployed it, silently reporting CSRF failures to a database.
```
I would change this to:

> and then soft-deployed it -- silently reporting CSRF failures to a database.

danielbeardsley committed:
Post: CSRF - more grammar and spelling fixes.

marczych commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+
+We've come up with a minimalist method of
+protecting our site from cross site request forgery ([CSRF][csrf]),
+and we'd like to share how we do it.
```
I don't really like the end of this sentence. Maybe splitting it into 2 sentences would be best? e.g. We've come up with a minimalist method of protecting our site from cross site request forgery ([CSRF][csrf]). Here's how we do it.

marczych commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+then soft-deployed it &mdash; silently reporting CSRF failures to a database.
+Once we'd worked out a few issues
+and were confident that we'd covered all our bases,
+we flipped the switch on enforcement
```
we flipped the **enforcement** switch on?

marczych commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+Once we'd worked out a few issues
+and were confident that we'd covered all our bases,
+we flipped the switch on enforcement
+and it's been working well since.
```
and it's been working well **ever** since.?

danielbeardsley replied:
Both are valid, I went with the terser, but it does sound like `ever` was unintentionally left out, I'll add it.

marczych commented on _posts/2013-07-21-csrf-how-we-do-it.md:
```diff
@@ -0,0 +1,91 @@
+and reduces the chance of mistakes.
+
+Here's the entire javascript implementation of our CSRF protection (Github [gist][gist]):
+{% gist 6060418 %}
```
The gist finds all forms and adds the csrf input when the page initially loads (`domready` event). Is it worth mentioning that something special needs to happen when dynamically creating forms in JS? (Something special needs to happen right?)

danielbeardsley replied:
Good point. I'll add that.

danielbeardsley committed:
Post: CSRF - improve first sentence

danielbeardsley committed:
Post: CSRF - Remove references to you / your

They are fine, but the post should remain consistent and it was easier
to remove a few yous than to add a bunch

danielbeardsley committed:
Post: CSRF - Add caveat about dynamic forms

Also, 'ever since' sounds better than 'since' though both are valid.

danielbeardsley commented:
Ok... I'm gonna merge this.
